libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -98,68 +117,160 @@ m_nameDispenser(_context.dispenser)
+		_functionDefinition.body = Block{_functionDefinition.location, move(memoryVariableInits)};
+		_functionDefinition.body.statements.emplace_back(ExpressionStatement{
+			_functionDefinition.location,
+			FunctionCall{
+				_functionDefinition.location,
+				Identifier{_functionDefinition.location, newFunctionName},
+				stackArguments | ranges::views::transform([&](TypedName const& _arg) {
+					return Expression{Identifier{
+						_functionDefinition.location,
+						_arg.name
+					}};
+				}) | ranges::to<vector<Expression>>
+			}
+		});
+		_functionDefinition.body.statements.emplace_back(Assignment{
+			_functionDefinition.location,
+			{Identifier{_functionDefinition.location, _functionDefinition.returnVariables.front().name}},
+			make_unique<Expression>(generateMemoryLoad(
+				m_context.dialect,
+				_functionDefinition.location,
+				*m_memoryOffsetTracker(_functionDefinition.returnVariables.front().name)
+			))
+		});
+		return;
+	}
+
+	if (!memoryVariableInits.empty())
+	{
+		memoryVariableInits += move(_functionDefinition.body.statements);
+		_functionDefinition.body.statements = move(memoryVariableInits);
+	}
+
+	_functionDefinition.returnVariables = _functionDefinition.returnVariables | ranges::views::filter([&](TypedName const& _name){
+		return !m_memoryOffsetTracker(_name.name);
+	}) | ranges::to<TypedNameList>;
 }
 
 void StackToMemoryMover::operator()(Block& _block)
 {
 	using OptionalStatements = std::optional<vector<Statement>>;
-	auto rewriteAssignmentOrVariableDeclaration = [&](
+
+	auto rewriteAssignmentOrVariableDeclarationLeftHandSide = [this](
 		auto& _stmt,
-		auto const& _variables
+		auto& _lhsVars
 	) -> OptionalStatements {
 		using StatementType = decay_t<decltype(_stmt)>;
-		if (_stmt.value)
-			visit(*_stmt.value);
-		bool leftHandSideNeedsMoving = util::contains_if(_variables, [&](auto const& var) {
-			return m_memoryOffsetTracker(var.name);
-		});
-		if (!leftHandSideNeedsMoving)
-			return {};
-
-		langutil::SourceLocation loc = _stmt.location;
 
-		if (_variables.size() == 1)
+		langutil::SourceLocation  loc = _stmt.location;
+		if (_lhsVars.size() == 1)
 		{
-			optional<YulString> offset = m_memoryOffsetTracker(_variables.front().name);
-			yulAssert(offset, "");
-			return generateMemoryStore(
-				m_context.dialect,
-				loc,
-				*offset,
-				_stmt.value ? *std::move(_stmt.value) : Literal{loc, LiteralKind::Number, "0"_yulstring, {}}
-			);
+			optional<YulString> offset = m_memoryOffsetTracker(_lhsVars.front().name);
+			if (offset)
+				return generateMemoryStore(
+					m_context.dialect,
+					loc,
+					*offset,
+					_stmt.value ? *std::move(_stmt.value) : Literal{loc, LiteralKind::Number, "0"_yulstring, {}}
+				);
+			else
+				return {};
 		}
+		FunctionCall const* functionCall = get_if<FunctionCall>(_stmt.value.get());
+		yulAssert(functionCall, "");
+		auto rhsSlots = m_functionReturnVariables.at(functionCall->functionName.name) |
```

Comment:
If `rhsSlots[i]` is non-nullopt, then that variable is moved to memory? Shouldn't it be called `rhsMemory` instead?

Reply:
(I'm assuming "slot" refers to "stack slots")

Reply:
"slots" are "memory slots" :-).